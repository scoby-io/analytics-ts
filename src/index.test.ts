import { Client } from './';
import nock from 'nock';

const USERAGENT = 'Mozilla/Safari/Chrome/1.2';
const IP_ADDRESS = '2.3.4.5';
const REFERRING_URL = 'https://www.referrer.de/i-refer?bar=foo';
const REQUESTED_URL =
  'www.this-host-was-called.com/and-this-was-the-path?with=query';

describe('Client', () => {
  beforeEach(() => {
    nock.cleanAll();
  });

  it('should initialize', () => {
    const client = new Client('xyz789');
    expect(client).toBeInstanceOf(Client);
  });

  it('should throw on error', async () => {
    const client = new Client('xyz789');

    nock('https://xyz789.s3y.io')
      .intercept(() => true, 'GET')
      .replyWithError('Internal Server Error');

    await expect(
      client.logPageView({
        ipAddress: IP_ADDRESS,
        requestedUrl: REQUESTED_URL,
        userAgent: USERAGENT,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should fail on non 204 response', async () => {
    const client = new Client('xyz789');

    nock('https://xyz789.s3y.io')
      .intercept(() => true, 'GET')
      .reply(201);

    await expect(
      client.logPageView({
        ipAddress: IP_ADDRESS,
        requestedUrl: REQUESTED_URL,
        userAgent: USERAGENT,
      }),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should log page view', async () => {
    const client = new Client('xyz789');

    const uris: string[] = [];
    nock('https://xyz789.s3y.io')
      .intercept((uri) => {
        uris.push(uri);
        return true;
      }, 'GET')
      .reply(204);

    await expect(
      client.logPageView({
        ipAddress: IP_ADDRESS,
        requestedUrl: REQUESTED_URL,
        userAgent: USERAGENT,
      }),
    ).resolves.toEqual(true);

    expect(uris).toMatchSnapshot();
  });

  it('should log page view with referrer', async () => {
    const client = new Client('xyz789');

    const uris: string[] = [];
    nock('https://xyz789.s3y.io')
      .intercept((uri) => {
        uris.push(uri);
        return true;
      }, 'GET')
      .reply(204);

    await expect(
      client.logPageView({
        ipAddress: IP_ADDRESS,
        requestedUrl: REQUESTED_URL,
        userAgent: USERAGENT,
        referringUrl: REFERRING_URL,
      }),
    ).resolves.toEqual(true);

    expect(uris).toMatchSnapshot();
  });

  it('should log page view with visitorId', async () => {
    const client = new Client('xyz789');

    const uris: string[] = [];
    nock('https://xyz789.s3y.io')
      .intercept((uri) => {
        uris.push(uri);
        return true;
      }, 'GET')
      .reply(204);

    await expect(
      client.logPageView({
        ipAddress: IP_ADDRESS,
        requestedUrl: REQUESTED_URL,
        userAgent: USERAGENT,
        visitorId: 'hello I am a value that wil be hashed before transfer',
      }),
    ).resolves.toEqual(true);

    expect(uris).toMatchSnapshot();
  });
});
