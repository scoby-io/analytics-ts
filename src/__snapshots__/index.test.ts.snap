// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client should fail on non 204 response 1`] = `"failed to log page viewfailed to log page view. Status code: 201"`;

exports[`Client should log page view 1`] = `
Array [
  "/count",
  "/count?vid=2dc4e4a6b96b8349edd6bad627dccee301389f0d92d528ce625f38a5819d7aca&ua=Mozilla%2FSafari%2FChrome%2F1.2&url=www.this-host-was-called.com%2Fand-this-was-the-path%3Fwith%3Dquery",
]
`;

exports[`Client should log page view with referrer 1`] = `
Array [
  "/count",
  "/count?vid=2dc4e4a6b96b8349edd6bad627dccee301389f0d92d528ce625f38a5819d7aca&ua=Mozilla%2FSafari%2FChrome%2F1.2&url=www.this-host-was-called.com%2Fand-this-was-the-path%3Fwith%3Dquery&ref=https%3A%2F%2Fwww.referrer.de%2Fi-refer%3Fbar%3Dfoo",
]
`;

exports[`Client should log page view with visitorId 1`] = `
Array [
  "/count",
  "/count?vid=cd8c117700c7d79bdd835dd6aa881750551d0c9e5c5bff4ca978b75715832df4&ua=Mozilla%2FSafari%2FChrome%2F1.2&url=www.this-host-was-called.com%2Fand-this-was-the-path%3Fwith%3Dquery",
]
`;

exports[`Client should throw on error 1`] = `"failed to log page viewInternal Server Error"`;
